### PR TITLE
Added generateId to SocketIO.Server.engine

### PR DESCRIPTION
**Pull request by sinan92**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://socket.io/docs/server-api/#server-engine-generateId>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/socket.io/index.d.ts [hunk 1]
```diff
@@ -81,7 +81,7 @@ interface SocketIOStatic {
 
 declare namespace SocketIO {
     interface Server {
-        engine: { ws: any };
+        engine: { ws: any, generateId: () => {} };
```
Comment (murbanowicz) [inline types/socket.io/index.d.ts:84]: this is not correct.
Based on the docs which you have provided the link to: 

1. takes the request as param
2. has to return a string
image

File: types/socket.io/v1/index.d.ts [hunk 2]
```diff
@@ -50,7 +50,7 @@ interface SocketIOStatic {
 
 declare namespace SocketIO {
     interface Server {
-        engine: { ws: any };
+        engine: { ws: any, generateId: () => {} };
```
Comment (murbanowicz) [inline types/socket.io/v1/index.d.ts:53]: same as with current version

**Review by murbanowicz — CHANGES_REQUESTED** [review]

File: types/socket.io/index.d.ts (hunk 1 above)
Comment (sinan92) [reply types/socket.io/index.d.ts]: My apologies you are right. I changed the code to be more compliant with the docs.

File: types/socket.io/v1/index.d.ts (hunk 2 above)
Comment (sinan92) [reply types/socket.io/v1/index.d.ts]: I'm not very sure what the "v1" folder is for. Do I need to leave this unchanged in my case?

File: types/socket.io/index.d.ts [hunk 3]
```diff
@@ -81,7 +81,7 @@ interface SocketIOStatic {
 
 declare namespace SocketIO {
     interface Server {
-        engine: { ws: any };
+        engine: { ws: any, generateId: (id: string) => string };
```
Comment (murbanowicz) [inline types/socket.io/index.d.ts:84]: > The function is called with a node request object (http.IncomingMessage) as the first parameter.

**Review by murbanowicz — CHANGES_REQUESTED** [review]
Please correct the type of the callback param

File: types/socket.io/index.d.ts (hunk 3 above)
Comment (sinan92) [reply types/socket.io/index.d.ts]: My apologies again I should read it better. I resolved it.

**Review by murbanowicz — APPROVED** [review]

**Comment by murbanowicz** [comment]
@sinan92 - please add some tests so we can merge it automatically ok?

**Comment by sinan92** [comment]
I never wrote tests for npm packages before but I hope I've done it right?

File: types/socket.io/socket.io-tests.ts [hunk 4]
```diff
@@ -235,3 +235,16 @@ function testServerEventEmitter() {
...
+function testOverwriteGenerateId() {
+    var io = socketIO.listen(80);
+    io.use((socket, next) => {
+        io.engine.generateId = () => {
+            return socket.handshake.query.deviceCode;
+        }
+        next();
+    })
+    .on('connection', (socket) => {
+        console.log(socket.id);
```
Comment (murbanowicz) [inline types/socket.io/socket.io-tests.ts:248]: please add something to throw if it is not as expected.

**Review by murbanowicz — CHANGES_REQUESTED** [review]

File: types/socket.io/socket.io-tests.ts (hunk 4 above)
Comment (sinan92) [reply types/socket.io/socket.io-tests.ts:248]: Ah okay

**Comment by murbanowicz** [comment]
@sinan92 I am not sure why I don't see your comments here :/ 
Please take a look at other tests here or in other packages. Should be straight forward.
It is not bad at the moment, but we should throw if it is not as expected so shouldn't be hard to improve?

**Comment by sinan92** [comment]
> @sinan92 I am not sure why I don't see your comments here :/
> Please take a look at other tests here or in other packages. Should be straight forward.
> It is not bad at the moment, but we should throw if it is not as expected so shouldn't be hard to improve?

I saw that the other tests in this package don't throw but only "console.log()" so I thought this was the way. But I think I got it right now.

File: types/socket.io/socket.io-tests.ts [hunk 5]
```diff
@@ -235,3 +235,20 @@ function testServerEventEmitter() {
...
+    io.use((socket, next) => {
+        io.engine.generateId = () => {
+            return hash;
+        }
+        next();
+    })
+    .on('connection', (socket) => {
+        console.log(socket.id);
+        if (socket.id !== hash) {
+            throw new Error("GenerateId has not been overwritten");
```
Comment (murbanowicz) [inline types/socket.io/socket.io-tests.ts:251]: just great!

File: types/socket.io/socket.io-tests.ts (hunk 4 above)
Comment (sinan92) [reply types/socket.io/socket.io-tests.ts:249]: Is this better?

Comment (murbanowicz) [reply types/socket.io/socket.io-tests.ts:249]: Waiting for the results of the test.

**Review by murbanowicz — APPROVED** [review]
Good job! Thank you :)

**Comment by murbanowicz** [comment]
@sinan92 - add required comment about ready to merge :)

**Comment by sinan92** [comment]
Ready to merge

**Pull request merged**